src/tls/server.c

```diff
@@ -0,0 +1,818 @@
+#endif
+    }
+
+  /* systemd socket activated? */
+  listen_fds = secure_getenv ("LISTEN_FDS");
+  if (listen_fds && check_sd_listen_pid ())
+    {
+      if (strcmp (listen_fds, "1") != 0)
+        errx (1, "This program can only accept exactly one socket from systemd, but got passed %s", listen_fds);
+      server->listen_fd = SD_LISTEN_FDS_START;
+      debug ("Server ready. Listening to systemd activated socket fd %i", server->listen_fd);
+    }
+  else
+    {
+      struct sockaddr_in sa_serv;
+      int optval = 1;
+
+      /* Listen to our port */
+      server->listen_fd = socket (AF_INET, SOCK_STREAM | SOCK_CLOEXEC, 0);
+      if (server->listen_fd < 0)
+        err (1, "failed to create server listening fd");
+
+      memset (&sa_serv, '\0', sizeof (sa_serv));
+      sa_serv.sin_family = AF_INET;
+      sa_serv.sin_addr.s_addr = INADDR_ANY;
+      sa_serv.sin_port = htons (port);
+
+      if (setsockopt (server->listen_fd, SOL_SOCKET, SO_REUSEADDR, (void *) &optval, sizeof (int)) < 0)
+        err (1, "failed to set socket option");
+      if (bind (server->listen_fd, (struct sockaddr *) &sa_serv, sizeof (sa_serv)) < 0)
+        err (1, "failed to bind to port %hu", port);
+      if (listen (server->listen_fd, 1024) < 0)
+        err (1, "failed to listen to server port");
+      debug ("Server ready. Listening on port %hu, fd %i", port, server->listen_fd);
+    }
+
+  /* epoll the listening fd */
+  server->epollfd = epoll_create1 (EPOLL_CLOEXEC);
+  if (server->epollfd < 0)
+    err (1, "Failed to create epoll fd");
+  ev.data.fd = server->listen_fd;
+  if (epoll_ctl (server->epollfd, EPOLL_CTL_ADD, server->listen_fd, &ev) < 0)
+    err (1, "Failed to epoll server listening fd");
+
+  return server;
+}
+
+void
+server_free (Server *server)
+{
+  close (server->epollfd);
+  close (server->listen_fd);
+
+  block_sigchld (1);
+
+  for (Connection *c = server->connections; c; )
+    {
+      Connection *cnext = c->next;
+      connection_free (c);
+      c = cnext;
+    }
+
+  for (WsInstance *ws = server->wss; ws; )
+    {
+      WsInstance *wsnext = ws->next;
+      ws_instance_free (ws);
+      ws = wsnext;
+    }
+  if (server->ws_notls)
+    ws_instance_free (server->ws_notls);
+
+  block_sigchld (0);
+
+  if (server->x509_cred)
+    {
+      gnutls_certificate_free_credentials (server->x509_cred);
+      gnutls_priority_deinit (server->priority_cache);
+    }
+
+  free (server);
+}
+
+/**
+ * server_poll_event: Wait for and process one event
+ *
+ * @timeout: number of milliseconds to wait for an event to happen; after that,
+ * the function will return. -1 will to block forever
+ *
+ * This can be any event on the listening socket, on connected client sockets,
+ * or from cockpit-ws children.
+ *
+ * Returns: 1 if the server should continue to run. 0 if the server has been
+ * idle for at least the idle_timeout specified in #server_new.
+ */
+int
+server_poll_event (Server *server, int timeout)
```

Comment:
No.  This is not okay.  Do not make a process that wakes up once a second to do nothing at all.  It should only wake up when there is actual work to do.